gslib/utils/copy_helper.py

```diff
@@ -663,58 +655,58 @@ def ConstructDstUrl(src_url,
-    #   gsutil mv gs://bucket/abcd gs://bucket/xyz
-    # when processing exp_src_url=gs://bucket/abcd/123
-    # exp_src_url_tail should become /123
-    # Note: mv.py code disallows wildcard specification of source URL.
-    recursive_move_to_new_subdir = True
-    exp_src_url_tail = (exp_src_url.url_string[len(src_url.url_string):])
-    dst_key_name = '%s/%s' % (exp_dst_url.object_name.rstrip('/'),
-                              exp_src_url_tail.strip('/'))
-
-  elif src_url_names_container and (exp_dst_url.IsCloudUrl() or
-                                    exp_dst_url.IsDirectory()):
-    # Case 2.  Container copy to a destination other than a file.
+  # Ignore the "multiple top-level sources" rule if using double wildcard **
+  # because that treats all files as top-level, in which case the user doesn't
+  # want to preserve directories.
+  preserve_src_top_level_dirs = '**' not in src_url.versionless_url_string and (
+      has_multiple_top_level_srcs or have_existing_dest_subdir)
+  if preserve_src_top_level_dirs or (src_url_names_container and
+                                     (exp_dst_url.IsCloudUrl() or
+                                      exp_dst_url.IsDirectory())):
+    # Case 1.  Container copy to a destination other than a file.
     # Build dst_key_name from subpath of exp_src_url past
     # where src_url ends. For example, for src_url=gs://bucket/ and
     # exp_src_url=gs://bucket/src_subdir/obj, dst_key_name should be
     # src_subdir/obj.
     src_url_path_sans_final_dir = GetPathBeforeFinalDir(src_url, exp_src_url)
-
     dst_key_name = exp_src_url.versionless_url_string[
         len(src_url_path_sans_final_dir):].lstrip(src_url.delim)
-    # Handle case where dst_url is a non-existent subdir.
-    if not have_existing_dest_subdir:
+
+    if not preserve_src_top_level_dirs:
+      # Only copy file name, not parent dir.
       dst_key_name = dst_key_name.partition(src_url.delim)[-1]
-    # Handle special case where src_url was a directory named with '.' or
-    # './', so that running a command like:
-    #   gsutil cp -r . gs://dest
-    # will produce obj names of the form gs://dest/abc instead of
-    # gs://dest/./abc.
+
     if dst_key_name.startswith('.%s' % os.sep):
+      # Handle special case where src_url was a directory named with '.' or
+      # './', so that running a command like:
+      #   gsutil cp -r . gs://dest
+      # will produce obj names of the form gs://dest/abc instead of
+      # gs://dest/./abc.
       dst_key_name = dst_key_name[2:]
-
   else:
-    # Case 3.
+    # Case 2.
     dst_key_name = exp_src_url.object_name.rpartition(src_url.delim)[-1]
 
-  if (not recursive_move_to_new_subdir and
-      (exp_dst_url.IsFileUrl() or _ShouldTreatDstUrlAsBucketSubDir(
-          have_multiple_srcs, exp_dst_url, have_existing_dest_subdir,
-          src_url_names_container, recursion_requested))):
+  if (exp_dst_url.IsFileUrl() or _ShouldTreatDstUrlAsBucketSubDir(
```

Comment:
Don't we have to handle special cases for `.` and `..` like we do in the `gcloud storage` code? Or is that handled already?

Reply:
Thanks

Reply:
Don't want to merge this PR with fixes for the `..` issue @rrauber is handling.

Reply:
That will lead to differences in the behavior for `..` upload case between gsutil and gcloud. Also, I think Ross is going to simply skip downloading any file that has `..` in the object name, which seems to be different than the upload case anyway.

Reply:
This is not addressed?

Reply:
Done